### PR TITLE
Ensure Level uses script-cache for relevant queries

### PR DESCRIPTION
**Pull request by wjordan**
This PR overloads `Level#find` and `Level#find_by` to return the script-cached objects instead of querying the database for each Level find.

Adds a `get_rubric` DB query test to prevent regressions on this new high-traffic route.

### CODE REVIEW

**Comment by sureshc** [comment]
Looks good to me!  Should we have someone in Learning Platform take a look so they can proactively include this approach in similar controller actions they implement in the future?

**Comment by wjordan** [comment]
> Should we have someone in Learning Platform take a look so they can proactively include this approach in similar controller actions they implement in the future?

In general sure, share the PR for others to take a look, no need to ask permission for this.

However, I wouldn't hold up this code as a great example to happily replicate with abandon. Our script-model-caching code has become a huge mess of hacks and patches and narrow workarounds on top of workarounds grown over the last five years. This PR arguably makes the whole mess worse with further workarounds targeting a specific set of query regressions, so I'd be very wary of duplicating it further unless absolutely necessary.

This system really demands someone (with advanced knowledge of Rails/ActiveRecord) spend a dedicated (and probably significant) chunk of time to unpack and refactor the caching logic into a manageable, modular component. Until that work can be prioritized/scheduled it'll continue to be an unholy mess sprawling throughout our application code.

**Review by sureshc — APPROVED** [review]

File: dashboard/app/models/script.rb [hunk 1]
```diff
@@ -405,7 +405,7 @@ def self.cache_find_level(level_identifier)
     # the db. Note the field trickery is to allow passing an ID as a string,
     # which some tests rely on (unsure about non-tests).
     field = level_identifier.to_i.to_s == level_identifier.to_s ? :id : :name
-    level = Level.find_by!(field => level_identifier)
+    level = Level.where(field => level_identifier).take!
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb:408]: why is this change necessary?

File: dashboard/app/models/levels/level.rb [hunk 2]
```diff
@@ -376,6 +376,23 @@ def calculate_ideal_level_source_id
     update_attribute(:ideal_level_source_id, ideal_level_source.id) if ideal_level_source
   end
 
+  # Ensure Level uses script-cache for relevant queries.
+  def self.find(id)
+    find_by(id: id)
+  end
+
+  # Ensure Level uses script-cache for relevant queries.
+  def self.find_by(opts)
```
Comment (davidsbailey) [inline dashboard/app/models/levels/level.rb:385]: Clever! I think there are a couple of assumptions being made here which could be worth noting in this method's comments:
1. in order to avoid an infinite loop, `Script.cache_find_level` must use `Level.find_by!`, not `Level.find_by`
2. in order to ensure Level uses script-cache for relevant queries, the rest of the application should use `Level.find_by`, not `Level.find_by!`

File: dashboard/app/models/levels/level.rb [hunk 3]
```diff
@@ -376,6 +376,23 @@ def calculate_ideal_level_source_id
...
+    other = opts.dup
+    id = other.delete(:id)&.to_i
+    name = other.delete(:name)
+    if (id || name) && other.empty?
+      Script.cache_find_level(id || name)
+    else
+      super(opts)
+    end
+  end
```
Comment (islemaster) [inline dashboard/app/models/levels/level.rb:394]: :cry: Overriding these ActiveRecord methods points out how often we're failing to think about perf.

Echoing your own comments here, if we're going in the direction of hiding caching logic behind an ActiveRecord-like interface, maybe we should go all-in and wrap the _whole_ query interface.  Maybe we can write a concern to do that.  [Maybe someone already has.](https://github.com/hooopo/second_level_cache)

I'll push this onto the list of tech debt for LP to consider in H1 2020.

File: dashboard/app/models/levels/level.rb [hunk 4]
```diff
@@ -376,6 +376,23 @@ def calculate_ideal_level_source_id
     update_attribute(:ideal_level_source_id, ideal_level_source.id) if ideal_level_source
   end
 
+  # Ensure Level uses script-cache for relevant queries.
+  def self.find(id)
+    find_by(id: id)
+  end
```
Comment (davidsbailey) [inline dashboard/app/models/levels/level.rb:382]: I am looking for usages of this, and I can't find any callers of this method: https://github.com/code-dot-org/code-dot-org/blob/b23da181be33b04abb3bba0b257919f123e55c8b/dashboard/app/models/script_level.rb#L78-L80

@Hamms any usage of `def level=` you're aware of? I couldn't find any. I could remove it in a subsequent PR, unless you can find any tricky ruby syntax that might still be calling it.

**Review by davidsbailey — APPROVED** [review]
Looks great to me, @wjordan !

File: dashboard/app/models/levels/level.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/models/levels/level.rb:394]: Good point @islemaster that this change as written sort of encourages not thinking about perf (except that you do have to think about it in order to call `.find_by` instead of `.find_by!`).

would you consider it an improvement to do a simple rename from all existing calls to `Level.find` to `Level.get_from_cache`, for example? and do a similar rename for `Level.find_by` --> `Level.find_by_from_cache`? That way, we would not have to remember as much in order to recognize new occurrences of ActiveRecord primitives as bypassing the cache.

@wjordan I would say please do not let this block your PR -- LP team can handle the cleanup.

File: dashboard/app/models/levels/level.rb (hunk 4 above)
Comment (Hamms) [reply dashboard/app/models/levels/level.rb:382]: Did you check not only for explicit calls to the method, but also assignments to the `level_id` property in general? My understanding is that the way ruby works, this method will be invoked if you simply execute `script_level.level_id = x`

Comment (davidsbailey) [reply dashboard/app/models/levels/level.rb:382]: yes, I looked for `level_id =` in instance methods of ScriptLevelsController and `.level_id =` globally.

Comment (Hamms) [reply dashboard/app/models/levels/level.rb:382]: there are no specific usages of this that I'm aware of; I'd be comfortable removing it

File: dashboard/app/models/levels/level.rb (hunk 3 above)
Comment (islemaster) [reply dashboard/app/models/levels/level.rb:394]: Yeah, I might - I've got this sense that I'd like us to override _all_ of the ActiveRecord query interface, or none of it, to make the behavior as unsurprising as possible.  Agreed on not blocking PR on this.

File: dashboard/app/models/levels/level.rb (hunk 4 above)
Comment (davidsbailey) [reply dashboard/app/models/levels/level.rb:382]: https://github.com/code-dot-org/code-dot-org/pull/31881
